include/xgboost/collective/result.h

```diff
@@ -0,0 +1,135 @@
+/**
+ *  Copyright 2023, XGBoost Contributors
+ */
+#pragma once
+
+#include <memory>   // for unique_ptr
+#include <sstream>  // for stringstream
+#include <string>   // for string
+#include <utility>  // for move
+
+namespace xgboost::collective {
+namespace detail {
+struct ResultImpl {
+  std::string message;
+  std::error_code errc{};  // optional for system error.
+
+  std::unique_ptr<ResultImpl> prev{nullptr};
+
+  ResultImpl() = delete;  // must initialize.
+  ResultImpl(ResultImpl const& that) = delete;
+  ResultImpl(ResultImpl&& that) = default;
+  ResultImpl& operator=(ResultImpl const& that) = delete;
+  ResultImpl& operator=(ResultImpl&& that) = default;
+
+  explicit ResultImpl(std::string msg) : message{std::move(msg)} {}
+  explicit ResultImpl(std::string msg, std::error_code errc)
+      : message{std::move(msg)}, errc{std::move(errc)} {}
+  explicit ResultImpl(std::string msg, std::unique_ptr<ResultImpl> prev)
+      : message{std::move(msg)}, prev{std::move(prev)} {}
+  explicit ResultImpl(std::string msg, std::error_code errc, std::unique_ptr<ResultImpl> prev)
+      : message{std::move(msg)}, errc{std::move(errc)}, prev{std::move(prev)} {}
+
+  [[nodiscard]] bool operator==(ResultImpl const& that) const {
+    if ((prev && !that.prev) || (!prev && that.prev)) {
+      // one of them doesn't have prev
+      return false;
+    }
+
+    auto cur_eq = message == that.message && errc == that.errc;
+    if (prev && that.prev) {
+      // recursive comparison
+      auto prev_eq = *prev == *that.prev;
+      return cur_eq && prev_eq;
+    }
+    return cur_eq;
+  }
+
+  [[nodiscard]] std::string Report() {
+    std::stringstream ss;
+    ss << "\n- " << this->message;
+    if (this->errc != std::error_code{}) {
+      ss << " system error:" << this->errc.message();
+    }
+
+    auto ptr = prev.get();
+    while (ptr) {
+      ss << "\n- ";
+      ss << ptr->message;
+
+      if (ptr->errc != std::error_code{}) {
+        ss << " " << ptr->errc.message();
+      }
+      ptr = ptr->prev.get();
+    }
+
+    return ss.str();
+  }
+};
+}  // namespace detail
+
+/**
+ * @brief An error type that's easier to handle than throwing dmlc exception. We can
+ *        record and propagate the system error code.
+ */
+struct Result {
```

Comment:
At some point in the future, we need to propagate the error to Python or other language bindings for delegating the error handling to higher-level frameworks like dask. At the moment, a functional form of error handling is easier to handle than exceptions.